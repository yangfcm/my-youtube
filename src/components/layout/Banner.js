import React from "react";
import { connect } from "react-redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Banner = ({ auth }) => {
  const { signedIn, user } = auth;
  // console.log(signedIn, user);
  return (
    <div className="d-flex align-items-center mb-2">
      <div style={{ height: "80px" }} className="d-flex align-items-center">
        {signedIn === null ? (
          ""
        ) : signedIn === true ? (
          <img
            style={{ borderRadius: "50%", maxHeight: "100%" }}
            src={user.avatar}
            alt={user.username}
          />
        ) : (
          <FontAwesomeIcon icon="user-astronaut" size="3x"></FontAwesomeIcon>
        )}
      </div>
      <h3 className="display-6 mx-3">
        {signedIn === null
          ? ""
          : signedIn === true
          ? `Welcome, ${user.username}`
          : "Welcome"}
      </h3>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
  };
};
export default connect(mapStateToProps)(Banner);
