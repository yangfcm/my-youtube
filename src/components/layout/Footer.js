import React from "react";

const Footer = () => {
  return (
    <div className="font-weight-light text-center bg-light py-3">
      My Youtube player developed by Fan Y.
    </div>
  );
};

export default Footer;
