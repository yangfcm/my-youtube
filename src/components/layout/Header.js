import React from "react";
import { connect } from "react-redux";
import { Link, withRouter } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import UserImage from "../common/UserImage";
import GoogleAuth from "../modules/GoogleAuth";

class Header extends React.Component {
  state = {
    showDropdownMenu: false,
    search: "",
    user: null,
  };

  handleSearch = (e) => {
    e.preventDefault();
    if (this.state.search.trim() === "") {
      return;
    }
    this.props.history.push(`/results?q=${this.state.search}`);
    this.setState({
      search: "",
    });
  };

  handleInput = (e) => {
    this.setState({
      search: e.target.value,
    });
  };

  handleSignin = () => {
    this.props.googleSignin();
  };

  render() {
    return (
      <nav className="navbar navbar-light bg-light mb-4">
        <Link to="/" className="navbar-brand">
          <FontAwesomeIcon
            icon="video"
            size="lg"
            style={{ color: "#FF0101" }}
          />
          <span className="font-weight-bold mx-2 d-none d-sm-inline">
            FanTube{" "}
          </span>
        </Link>
        <div className="flex-grow-1 d-flex align-items-center justify-content-end">
          <form className="d-flex" onSubmit={this.handleSearch}>
            <input
              type="text"
              placeholder="Search video..."
              className="form-control"
              name="search"
              value={this.state.search}
              onChange={this.handleInput}
            />
            <button className="btn btn-outline-dark mx-2" type="submit">
              <FontAwesomeIcon icon="search" />
            </button>
          </form>
          {this.props.auth.signedIn === true && (
            <UserImage user={this.props.auth.user} />
          )}
          {this.props.auth.signedIn === false && (
            <div
              style={{ height: "55px" }}
              className="d-flex align-items-center"
            >
              <button className="btn btn-primary" onClick={this.handleSignin}>
                <FontAwesomeIcon icon="sign-in-alt" />
                {"   "}
                Sign In
              </button>
            </div>
          )}
        </div>
      </nav>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
  };
};

export default connect(mapStateToProps)(withRouter(GoogleAuth(Header)));
