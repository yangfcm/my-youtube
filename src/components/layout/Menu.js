import React from "react";
import { NavLink } from "react-router-dom";

const Menu = ({ menuItems = [] }) => {
  return (
    <div>
      <nav className="nav nav-tabs nav-fill">
        {menuItems.map(item => {
          return (
            <NavLink
              key={item.name}
              to={item.link}
              exact={item.exact}
              className="nav-item nav-link"
              activeClassName="active"
            >
              {item.name}
            </NavLink>
          );
        })}
      </nav>
    </div>
  );
};

export default Menu;
