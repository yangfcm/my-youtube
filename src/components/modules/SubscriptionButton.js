import React from "react";
import { connect } from "react-redux";
import { fetchChannelSubscription } from "../../actions/app";
import { subscribeChannel, unsubscribeChannel } from "../../actions/app";

class SubscriptionButton extends React.Component {
  state = {
    isSubscribed: null,
    isLoading: false,
    buttonText: "",
  };

  componentDidMount = async () => {
    // console.log(this.props.auth);
    if (!this.props.auth.signedIn) return;
    this.fetchSubscriptionStatus();
  };

  componentDidUpdate = (prevProps) => {
    if (!prevProps.auth.signedIn && this.props.auth.signedIn) {
      this.fetchSubscriptionStatus();
    }
    if (prevProps.auth.signedIn && !this.props.auth.signedIn) {
      this.setState({
        isSubscribed: null,
      });
    }
  };

  fetchSubscriptionStatus = async () => {
    const accessToken = localStorage.getItem("access_token");
    await this.props.fetchChannelSubscription(
      this.props.channelId,
      accessToken
    );
    this.setState({
      isSubscribed: this.props.channel.isSubscribed,
      buttonText: this.props.channel.isSubscribed
        ? "Subscribed"
        : "Unsubscribed",
    });
  };

  handleSubscribe = async () => {
    const accessToken = localStorage.getItem("access_token");
    this.setState({
      isLoading: true,
    });
    try {
      await this.props.subscribeChannel(this.props.channelId, accessToken);
      // console.log(this.props.subscription);
      if (
        this.props.subscription &&
        this.props.subscription.snippet.resourceId.channelId ===
          this.props.channelId
      ) {
        this.setState({
          isSubscribed: true,
          buttonText: "Subscribed",
        });
      }
    } finally {
      this.setState({
        isLoading: false,
      });
    }
  };

  handleUnsubscribe = async () => {
    const accessToken = localStorage.getItem("access_token");
    this.setState({
      isLoading: true,
    });
    try {
      await this.props.unsubscribeChannel(this.props.channelId, accessToken);
      if (!this.props.subscription) {
        this.setState({
          isSubscribed: false,
          buttonText: "Unsubscribed",
        });
      }
    } finally {
      this.setState({
        isLoading: false,
      });
    }
  };

  handleToggleButtonTextOnHover = () => {
    if (this.state.buttonText === "Subscribed") {
      this.setState({
        buttonText: "Unsubscribe",
      });
    }
    if (this.state.buttonText === "Unsubscribed") {
      this.setState({
        buttonText: "Subscribe",
      });
    }
  };

  handleToggleButtonTextOnLeave = () => {
    if (this.state.isSubscribed === true) {
      this.setState({
        buttonText: "Subscribed",
      });
    } else {
      this.setState({
        buttonText: "Unsubscribed",
      });
    }
  };

  render() {
    const { isSubscribed, isLoading, buttonText } = this.state;
    return isSubscribed === null ? (
      " "
    ) : (
      <button
        className={`btn ${
          isSubscribed ? "btn-outline-danger" : "btn-outline-dark"
        }`}
        onMouseOver={this.handleToggleButtonTextOnHover}
        onMouseLeave={this.handleToggleButtonTextOnLeave}
        onClick={isSubscribed ? this.handleUnsubscribe : this.handleSubscribe}
        style={{ width: "120px" }}
        disabled={isLoading}
      >
        {buttonText}
      </button>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
    channel: state.channel,
    subscription: state.subscription,
  };
};
export default connect(mapStateToProps, {
  fetchChannelSubscription,
  subscribeChannel,
  unsubscribeChannel,
})(SubscriptionButton);
