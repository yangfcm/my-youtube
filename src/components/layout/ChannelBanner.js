import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { fetchChannelIntro } from "../../actions/channel";
import SubscriptionButton from "../modules/SubscriptionButton";

class ChannelBanner extends React.Component {
  state = {
    channelIntro: null,
  };
  componentDidMount = async () => {
    const { channelId } = this.props;
    await this.props.fetchChannelIntro(channelId);
    if (this.props.channel) {
      if (this.props.channel.items && this.props.channel.items[0]) {
        this.setState({
          channelIntro: this.props.channel.items[0],
        });
      } else {
        this.props.history.push("/page-not-found");
      }
    }
  };

  render() {
    const { channelIntro } = this.state;
    return (
      channelIntro && (
        <div className="d-flex align-items-center mb-2">
          <div style={{ height: "80px" }} className="d-flex align-items-center">
            {" "}
            <img
              style={{ borderRadius: "50%", maxHeight: "100%" }}
              src={channelIntro.snippet.thumbnails.medium.url}
              alt={channelIntro.snippet.title}
            />
          </div>
          <div className="flex-grow-1 d-flex justify-content-between align-items-center flex-wrap">
            <h3 className="display-6 mx-3 text-primary font-weight-bold ">
              {channelIntro.snippet.title}
            </h3>
            <div className="flex-grow-1 d-flex justify-content-end">
              <SubscriptionButton channelId={this.props.channelId} />
            </div>
          </div>
        </div>
      )
    );
  }
}

const mapStateToProps = (state) => {
  return {
    channel: state.channel.channelIntro,
  };
};
export default connect(mapStateToProps, {
  fetchChannelIntro,
})(withRouter(ChannelBanner));
