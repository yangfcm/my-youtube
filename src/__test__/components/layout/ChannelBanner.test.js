import React from "react";
// import { Provider } from "react-redux";
// import { BrowserRouter as Router, Link, withRouter } from "react-router-dom";
import { shallow } from "enzyme";
// import configMockStore from "redux-mock-store";
import { ChannelBanner } from "components/layout/ChannelBanner";
import SubscriptionButton from "components/modules/SubscriptionButton";
import {
  channelIntroData,
  channelIntroEmptyData,
} from "__test__/fixtures/channel";

describe("Test ChannelBanner component", () => {
  let wrapper;
  const fetchChannelIntro = jest.fn();
  const mockPush = jest.fn();

  beforeEach(() => {
    // const mockStore = configMockStore();
    // store = mockStore({
    //   channel: { channelIntro: channelIntroData },
    // });
    // store.dispatch = jest.fn();
    // wrapper = mount(
    //   <Provider store={store}>
    //     <Router>
    //       <ChannelBanner />
    //     </Router>
    //   </Provider>
    // );
    wrapper = shallow(
      <ChannelBanner
        fetchChannelIntro={fetchChannelIntro}
        history={{
          push: mockPush,
        }}
        channelId={channelIntroData.items[0].id}
      />
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("should call fetchChannelIntro", () => {
    expect(fetchChannelIntro).toHaveBeenCalled();
  });

  describe("If channelIntro is fetched", () => {
    beforeEach(() => {
      wrapper.setState({
        channelIntro: channelIntroData.items[0],
      });
    });

    it("should render component ", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render correct channel image ", () => {
      const image = wrapper.find("img");
      expect(image.prop("src")).toBe(
        channelIntroData.items[0].snippet.thumbnails.medium.url
      );
      expect(image.prop("alt")).toBe(channelIntroData.items[0].snippet.title);
    });

    it("should render channel title ", () => {
      expect(wrapper.find("h3").text()).toContain(
        channelIntroData.items[0].snippet.title
      );
    });

    it("should render Subscription button and pass channel id to it", () => {
      const subscriptionButton = wrapper.find(SubscriptionButton);
      expect(subscriptionButton.exists()).toBe(true);
      expect(subscriptionButton.prop("channelId")).toBe(
        channelIntroData.items[0].id
      );
    });
  });

  describe("If channel not found", () => {
    beforeEach(() => {
      wrapper.setProps({
        channel: channelIntroEmptyData,
      });
    });

    it("should navigate to not-found page", () => {
      expect(mockPush).toHaveBeenCalled();
    });
  });
});
