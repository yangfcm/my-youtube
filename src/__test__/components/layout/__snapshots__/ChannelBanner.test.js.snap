// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChannelBanner component If channelIntro is fetched should render component  1`] = `
<div
  className="d-flex align-items-center mb-2"
>
  <div
    className="d-flex align-items-center"
    style={
      Object {
        "height": "80px",
      }
    }
  >
     
    <img
      alt="Mark Rober"
      src="https://yt3.ggpht.com/a/AATXAJxOeBwKmjYvfuLwFuTuD3zFTffWCZbT5b5ufr8=s240-c-k-c0xffffffff-no-rj-mo"
      style={
        Object {
          "borderRadius": "50%",
          "maxHeight": "100%",
        }
      }
    />
  </div>
  <div
    className="flex-grow-1 d-flex justify-content-between align-items-center flex-wrap"
  >
    <h3
      className="display-6 mx-3 text-primary font-weight-bold "
    >
      Mark Rober
    </h3>
    <div
      className="flex-grow-1 d-flex justify-content-end"
    >
      <Connect(SubscriptionButton)
        channelId="UCY1kMZp36IQSyNx_9h4mpCg"
      />
    </div>
  </div>
</div>
`;
